container-chains/templates/frontier/node/src/rpc/mod.rs

```diff
@@ -0,0 +1,330 @@
+            )
+            .into_rpc(),
+        )?;
+    }
+
+    io.merge(
+        Net::new(
+            Arc::clone(&client),
+            network.clone(),
+            // Whether to format the `peer_count` response as Hex (default) or not.
+            true,
+        )
+        .into_rpc(),
+    )?;
+
+    io.merge(Web3::new(Arc::clone(&client)).into_rpc())?;
+    io.merge(
+        EthPubSub::new(
+            pool,
+            Arc::clone(&client),
+            sync.clone(),
+            subscription_task_executor,
+            overrides,
+            pubsub_notification_sinks,
+        )
+        .into_rpc(),
+    )?;
+
+    Ok(io)
+}
+
+pub struct SpawnTasksParams<'a, B: BlockT, C, BE> {
+    pub task_manager: &'a TaskManager,
+    pub client: Arc<C>,
+    pub substrate_backend: Arc<BE>,
+    pub frontier_backend: Arc<fc_db::Backend<B>>,
+    pub filter_pool: Option<FilterPool>,
+    pub overrides: Arc<OverrideHandle<B>>,
+    pub fee_history_limit: u64,
+    pub fee_history_cache: FeeHistoryCache,
+    /// Chain syncing service
+    pub sync_service: Arc<SyncingService<B>>,
+    /// Chain syncing service
+    pub pubsub_notification_sinks: Arc<
+        fc_mapping_sync::EthereumBlockNotificationSinks<
+            fc_mapping_sync::EthereumBlockNotification<B>,
+        >,
+    >,
+}
+
+use fc_mapping_sync::{MappingSyncWorker, SyncStrategy};
+/// Spawn the tasks that are required to run Moonbeam.
+pub fn spawn_essential_tasks<B, C, BE>(params: SpawnTasksParams<B, C, BE>)
+where
+    C: ProvideRuntimeApi<B> + BlockOf,
+    C: HeaderBackend<B> + HeaderMetadata<B, Error = BlockChainError> + 'static,
+    C: BlockchainEvents<B> + StorageProvider<B, BE>,
+    C: Send + Sync + 'static,
+    C::Api: EthereumRuntimeRPCApi<B>,
+    C::Api: BlockBuilder<B>,
+    B: BlockT<Hash = H256> + Send + Sync + 'static,
+    B::Header: HeaderT<Number = u32>,
+    BE: Backend<B> + 'static,
+    BE::State: StateBackend<BlakeTwo256>,
+{
+    // Frontier offchain DB task. Essential.
+    // Maps emulated ethereum data to substrate native data.
+    params.task_manager.spawn_essential_handle().spawn(
+        "frontier-mapping-sync-worker",
+        Some("frontier"),
+        MappingSyncWorker::new(
+            params.client.import_notification_stream(),
+            Duration::new(6, 0),
+            params.client.clone(),
+            params.substrate_backend.clone(),
+            params.overrides.clone(),
+            params.frontier_backend.clone(),
+            3,
+            0,
+            SyncStrategy::Parachain,
+            params.sync_service.clone(),
+            params.pubsub_notification_sinks.clone(),
+        )
+        .for_each(|()| futures::future::ready(())),
+    );
+
+    // Frontier `EthFilterApi` maintenance.
+    // Manages the pool of user-created Filters.
+    if let Some(filter_pool) = params.filter_pool {
+        // Each filter is allowed to stay in the pool for 100 blocks.
+        const FILTER_RETAIN_THRESHOLD: u64 = 100;
```

Comment:
Makes me wonder how this (and other block-based on-memory caches) will work with parathreads. Are those retained in memory for an undetermined amount of time for X blocks?

Reply:
good question. I guess blocks makes more sense

Reply:
let's review later, will add TODO for this